chore: add cypress

JacobMGEvans opened the pull request:
**5/24/20**
- initial setup, current test assert true is true
- next step is creating mock data for fields and utilizing the inputs

**5/25/20**
- Added data-cyp on the suggested query in component QueryAdvise
- Added comments to seed file (current unused mostly)
- Added two tests for smoke testing app main functionality

**5/29/20**
dda125b
- Updates to ESlint merged
- Updates to package.json merged
- Change in lock incoming change added with current

72eac23
- Fixed Linter Errors -- unused variables

960a767
- Tests cover the functionality of query options
- Clearing and user input of the code fields
- Clicking on the HTML previews
- Validating the query option selections change the descriptive content

smeijer commented:
Nice! Can you let me know if/when this is ready for review?

JacobMGEvans commented:
Yeah, that stuff is auto-generated by cypress, I tried removing it but it seems if to replace it if there isn't something there.
 I plan to continue working on this but figured it was a good initial point, so I would say it's ready as a starting point.

smeijer commented:
Deleting is possible, but requires you to add an empty config file.

Just create an `cypress.json` in the project root, containing nothing more than `{}`. Then you can delete those files :)

JacobMGEvans commented:
Yeah, but I will end up using fixtures soon anyways, this was mostly an initial setup with a smoke test.

JacobMGEvans commented:
@smeijer Updates:
- Added `fixtures`
- Refactored 
- Abstracted reusable functions to `support`
- Improvements to some of the testing (less brittle hopefully)

smeijer commented:
Thanks :+1:  Just wanted to let you know that I've seen it. I need a bit more time for a good check though. I'll come back to this.

JacobMGEvans commented:
I will have to resolve recent conflicts or someone else can, I am good either way. I won't be able to get to it tonight most likely.

JacobMGEvans commented:
> I will have to resolve recent conflicts or someone else can, I am good either way. I won't be able to get to it tonight most likely.

@smeijer I resolved the conflicts 😄

smeijer commented:
Awesome! Thanks. I'll try to take a look at it this weekend.

MichaelDeBoey commented on cypress.json:
```diff
@@ -0,0 +1,5 @@
+{
+        "baseUrl": "http://localhost:1234",
```
I think you need to run `prettier` again here 🤔

JacobMGEvans replied:
Isn't the husky hook running the linting and prettier application wide?

JacobMGEvans replied:
I checked it is not running on Husky hooks and apparently the root files are being ignored by Husky.

smeijer commented:
@JacobMGEvans, Thanks for your work on this one. I've taken a look at it, and have a few questions if you don't mind.

I'm missing a bit of information about where we are, and where we are going to. Can you help me out here? 

You're listing a changelog in your PR comment, but that we can see on the `commits` tab :innocent: 

I'd rather see:

- What do we currently have?
  - cypres test setup
  - cases for:
    - ...

- When is this PR done? And what will we have then?

- What's the added value of having cypres, over the current jest setup?

JacobMGEvans commented:
> @JacobMGEvans, Thanks for your work on this one. I've taken a look at it, and have a few questions if you don't mind.
> 
> I'm missing a bit of information about where we are, and where we are going to. Can you help me out here?
> 
> You're listing a changelog in your PR comment, but that we can see on the `commits` tab 😇
> 
> I'd rather see:
> 
> * What do we currently have?
>   
>   * cypres test setup
>   * cases for:
>     
>     * ...
> * When is this PR done? And what will we have then?
This PR can be considered finished, it's a lofty initial setup for tests. The commits I mentioned in the PR are to tell you what you will have... 
> * What's the added value of having cypres, over the current jest setup?
The value would be E2E, which I asked in #7 if Cypress was something interested in being integrated to the project.

smeijer commented:
I didn't mean it in a bad way. I still believe that it's a good thing to have tests. Either jest or cypress, or both.

Looking at the tests present in this PR, I see one fundamental issue. A lot of tests are based on either custom data (`data-cy`) attributes, or on class names.

For example:

```js
  it('Gives a suggested query more specific than the used data-testid', () => {
    cy.get('[data-cy=suggested-query]').should(
      'contain',
      "screen.getByText('Hello World')",
    );

    cy.get('.query-editor > .w-full > .CodeMirror').should(
      'contain',
      "screen.getByTestId('hello')",
    );
  });
```

As soon as we decide to remove the `w-full` class, this test will fail. I don't believe that's how we should structure our tests.

That's the opposite of one of the primary goals of this testing-playground. _"Get the best selector available. If that's not possible, improve your markup."_

JacobMGEvans commented:
Gotcha I edited the previous comment for more details as well.

As for the `data-cy`, that was actually the Cypress docs suggestion, they aren't using RTL under the hood its jQuery I believe. So there would need to be markup improvements if Cypress is expected to use certain selectors, overall it would be a good example of refactoring from Green -> Red -> Green if the markup needs updating.

smeijer commented:
> Gotcha I edited the previous comment for more details as well.
> 
> As for the `data-cy`, that was actually the Cypress docs suggestion, they aren't using RTL under the hood its jQuery I believe.

See (and install?) https://testing-library.com/docs/cypress-testing-library/intro

JacobMGEvans commented:
> > Gotcha I edited the previous comment for more details as well.
> > As for the `data-cy`, that was actually the Cypress docs suggestion, they aren't using RTL under the hood its jQuery I believe.
> 
> See (and install?) https://testing-library.com/docs/cypress-testing-library/intro

https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements

JacobMGEvans commented:
I can definitely work on that next! 😄

JacobMGEvans commented:
I was definitely waist-deep in Cypress docs and totally forgot about Testing Libs Cypress Library! lmao duh!

JacobMGEvans commented:
@smeijer Would you like that implemented in this PR before it goes in or as a next milestone for the Cypress integration?

smeijer commented:
I would prefer to have it in this PR, before we merge.

The codebase is moving quite fast, as the first thing I've shipped was more of a prototype level, than a production-grade thing.

Because of that, a lot of things are changing. ClassNames is one of those things. If we would merge this pull-request in it's current state, I'm sure tests will break all the time. 

And then it would slow us down, instead of giving assurance that everything still works as it should be.

smeijer commented:
Any wise words on what to do with this PR? I would love to have some e2e tests, but I have no Idea how much work it would be to bring this PR to a workable state.

ddehart commented:
@smeijer I figured I'd take a stab at this as a part of learning more about testing JavaScript, and the markup -- specifically the implementation of `codemirror` -- makes it challenging to make use of the principles of Testing Library.

Acknowledging that I'm a complete novice, attempting something even as simple as selecting the editor to clear it out and type something as a user would may be impossible without brittle selectors or acrobatics like grabbing parents of parents. I even went so far as trying to use Testing Playground on Testing Playground to identify good selectors without much luck. :joy: 

Someone with a bunch more Cypress/Testing Library experience can probably find an elegant solution, but I couldn't even really get started without considering class selectors or thinking about injecting a bunch of `testid`s into the markup.

I hope that perspective helps as this moves forward!

smeijer commented:
> Someone with a bunch more Cypress/Testing Library experience can probably find an elegant solution, but I couldn't even really get started without considering class selectors or thinking about injecting a bunch of `testid`s into the markup.

If using testIds or classNames is the only thing that can make this work (easily), then we should do that.

Does that mean that this branch works when we rebase it? We currently don't have many tests, so I'm eager to get this one merged.

ddehart commented:
> If using testIds or classNames is the only thing that can make this work (easily), then we should do that.

How would you feel about having actual `id`s associated with the editable components? That would make selection easy, potentially less brittle than classnames, and not so specific to a testing framework.

JacobMGEvans commented:
> > If using testIds or classNames is the only thing that can make this work (easily), then we should do that.
> 
> How would you feel about having actual `id`s associated with the editable components? That would make selection easy, potentially less brittle than classnames, and not so specific to a testing framework.

I think as long as it's utilizing the https://testing-library.com/docs/cypress-testing-library/intro/ and following the accessibility priority, then IDs necessity should be minimized. I never got around to that, unfortunately.

ddehart commented:
> I think as long as it's utilizing the https://testing-library.com/docs/cypress-testing-library/intro/

Oh, sorry, I was referring specifically to the HTML `id` attribute as opposed to a custom `testid` attribute. I'm just curious if HTML `id`s would be preferred for this project over `testid`s

smeijer commented:
`testIds` and `id` serve the same purpose in this regard. So in that case, I would say that a `testId` is preferable, as it at least makes its intension clear.

ddehart commented:
@JacobMGEvans do you plan on rebasing with your changes? If not, I can take a stab at it.

marcosvega91 commented:
This PR is very old, so I think that is better to start a new PR and copy what @JacobMGEvans has done here in the new one ? What do you think guys?

smeijer commented:
That's what a rebase does 😊

JacobMGEvans commented:
> @JacobMGEvans do you plan on rebasing with your changes? If not, I can take a stab at it.

You can use this branch if you like or create your own from scratch. I had a lot of the setup done which was most of the pain. I think I needed to add https://testing-library.com/docs/cypress-testing-library/intro/ (CTL)

The current tests were mostly smoke tests, just needed that CTL, however, if you DO do it from scratch keep this in mind because it was a big hurdle with the `CodeMirror` 
https://stackoverflow.com/questions/62012319/how-can-i-clear-a-codemirror-editor-field-from-cypress

JacobMGEvans commented:
I'll try to update the PR with upstream main, RN

smeijer commented:
The conflicting files doesn't seem to be much. Does it make sense to resolve the conflict, merge, and improve in a new PR?

JacobMGEvans commented:
Well, that is odd... The branch and work have disappeared. I know I didn't delete it.

JacobMGEvans commented:
I will figure this out later.

smeijer commented:
@JacobMGEvans, can it be because origin has changed in the mean time? It moved from `smeijer/testing-playground` to `testing-library/testing-playground`.

JacobMGEvans commented:
> @JacobMGEvans, can it be because origin has changed in the mean time? It moved from `smeijer/testing-playground` to `testing-library/testing-playground`.

Very possible 😅 !  The repo was originally forked from that old one. I don't know what the effects would normally be but explains the "unknown repository." I also rebased already and the work might have been `develop` local  I can try to undo that or find the changes in a detached `HEAD` if not the PR at least has most of the files changes here

smeijer commented:
Superseeded by https://github.com/testing-library/testing-playground/pull/313#issuecomment-739971737. Again, I'm very sorry that I haven't merged this sooner.
